### PR TITLE
Add nonlinear A_s option and fix normalization bugs

### PR DESCRIPTION
**Pull request by justjhong**
The baseline distances assume u is being sampled rather than just the mean being taken for each counterfactual. So, to use normalized distances, we must take a monte carlo sample rather than taking the mean for the latent representations.

### CODE REVIEW

File: src/scvi_v2/_model.py [hunk 1]
```diff
@@ -251,24 +257,30 @@ def inference_fn(
...
                 )
             else:
-                per_sample_inference_fn = lambda cf_sample: inference_fn(x, sample_index, cf_sample)
-                return jax.lax.transpose(jax.lax.map(per_sample_inference_fn, cf_sample), (1, 0, 2))
+
+                def per_sample_inference_fn(pair):
+                    rngs, cf_sample = pair
+                    return inference_fn(rngs, x, sample_index, cf_sample)
+
+                return jax.lax.transpose(jax.lax.map(per_sample_inference_fn, (rngs, cf_sample)), (1, 0, 2))
```
Comment (PierreBoyeau) [inline src/scvi_v2/_model.py]: it looks like here `stacked_rngs` and `rngs` should denote the same thing right?

File: src/scvi_v2/_model.py [hunk 2]
```diff
@@ -277,7 +289,14 @@ def mapped_inference_fn(
             inputs = self.module._get_inference_input(
                 array_dict,
             )
+            # Generate a set of RNGs for every cf_sample.
+            rngs_list = [self.module.rngs for _ in range(cf_sample.shape[0])]
+            stacked_rngs = {
+                required_rng: jnp.concatenate([rngs_dict[required_rng][None] for rngs_dict in rngs_list], axis=0)
+                for required_rng in self.module.required_rngs
+            }
```
Comment (PierreBoyeau) [inline src/scvi_v2/_model.py:298]: could you add a comment here?

File: tests/test_model.py [hunk 3]
```diff
@@ -96,11 +95,52 @@ def test_mrvi():
     # tests __repr__
     print(model)
 
+def test_mrvi_nonlinear():
+    adata = synthetic_iid()
```
Comment (PierreBoyeau) [inline tests/test_model.py:100]: can you pre-commit?

File: src/scvi_v2/_components.py [hunk 4]
```diff
@@ -50,6 +50,26 @@ def __call__(self, inputs: NdArray, training: Optional[bool] = None) -> NdArray:
...
+    @nn.compact
+    def __call__(self, inputs: NdArray, training: Optional[bool] = None) -> dist.Normal:  # noqa: D102
+        training = nn.merge_param("training", self.training, training)
+        h = inputs
+        for _ in range(self.n_layers):
+            h = ResnetBlock(
+                n_out=self.n_hidden, internal_activation=self.activation, output_activation=self.activation
+            )(h, training=training)
+        return Dense(self.n_out)(h)
+
```
Comment (PierreBoyeau) [inline src/scvi_v2/_components.py:71]: have you fixed the sparsity issue? It looks to me that the last units get a `relu` activation still.

File: src/scvi_v2/_model.py [hunk 5]
```diff
@@ -394,36 +421,41 @@ def _compute_local_baseline_dists(
...
+            return A_s.reshape(sample_covariate.shape[0], module.pz.n_latent, module.pz.n_latent)
+
+        def apply_get_A_s(u, sample_covariate):
             vars_in = {"params": self.module.params, **self.module.state}
             rngs = self.module.rngs
-            A_s = self.module.apply(vars_in, rngs=rngs, method=get_A_s, sample_index=sample_index).reshape(
-                sample_index.shape[0], self.module.n_latent, self.module.n_latent
-            )
+            A_s = self.module.apply(vars_in, rngs=rngs, method=get_A_s, u=u, sample_covariate=sample_covariate)
             return A_s
```
Comment (PierreBoyeau) [inline src/scvi_v2/_model.py:441]: at which point here do you use MC sampling instead of closed form? Im a bit lost

**Review by PierreBoyeau — APPROVED** [review]
There are a few things I don't understand, but it looks great to me!!

File: src/scvi_v2/_components.py (hunk 4 above)
Comment (justjhong) [reply src/scvi_v2/_components.py:71]: Dense is purely linear I believe, no activation

Comment (justjhong) [reply src/scvi_v2/_components.py:71]: double checked and i get real outputs!

File: src/scvi_v2/_model.py (hunk 1 above)
Comment (justjhong) [reply src/scvi_v2/_model.py]: yep ill rename

File: tests/test_model.py (hunk 3 above)
Comment (justjhong) [reply tests/test_model.py:100]: whats the issue here? precommit doesn't complain

File: src/scvi_v2/_model.py (hunk 5 above)
Comment (justjhong) [reply src/scvi_v2/_model.py:441]: It's all closed form except for the final step where the eigenvalues get multiplied with chisquare samples
